FinSearchUnified/Bundle/SearchBundleFindologic/FacetHandler/CategoryFacetHandler.php

```diff
@@ -0,0 +1,194 @@
+    {
+        foreach ($actives as $active) {
+            list($parent, $child) = explode('_', $active);
+
+            // Create structured array and recursively create category tree
+            $categories[$parent] = [
+                'active' => $child ? false : true,
+                'children' => $child ? self::prepareCategoryTree([$child], $categories) : []
+            ];
+        }
+
+        return $categories;
+    }
+
+    /**
+     * Parse the filter items and build an array structure
+     *
+     * @param SimpleXMLElement $filterItems
+     * @param array $actives
+     *
+     * @return array
+     */
+    private function parseCategories(SimpleXMLElement $filterItems, array $actives)
+    {
+        $categories = [];
+
+        foreach ($filterItems as $filterItem) {
+            $name = (string)$filterItem->name;
+            $frequency = (int)$filterItem->frequency;
+
+            // If category is in actives array, then set active to true
+            // And recursively parse child categories
+            $isActive = $this->keyExists($name, $actives);
+            $categories[$name] = [
+                'active' => $isActive,
+                'children' => $filterItem->items->item ? self::parseCategories($filterItem->items->item, $actives) : []
+            ];
+
+            if ($frequency) {
+                $categories[$name]['frequency'] = $frequency;
+            }
+        }
+
+        return $categories;
+    }
+
+    /**
+     * Convert the array of categories to an array of tree items
+     *
+     * @param array $categories
+     * @param string|null $parent
+     *
+     * @return array
+     */
+    private function getTreeItems(array $categories, $parent = null)
+    {
+        $items = [];
+
+        foreach ($categories as $key => $category) {
+            if ($parent !== null) {
+                $id = $parent . '_' . $key;
+            } else {
+                $id = $key;
+            }
+
+            $label = $key;
+
+            if (array_key_exists('frequency', $category) && !$category['active']) {
+                $label = sprintf('%s (%d)', $key, $category['frequency']);
+            }
+
+            $item = new TreeItem(
+                $id,
+                $label,
+                $category['active'],
+                self::getTreeItems($category['children'], $id)
+            );
+
+            $items[] = $item;
+        }
+
+        return $items;
+    }
+
+    /**
+     * Helper method to check if needle exists in array recursively
+     *
+     * @param mixed $needle
+     * @param array $haystack
+     *
+     * @return bool
+     */
+    private function keyExists($needle, array $haystack)
+    {
+        foreach ($haystack as $key => $value) {
+            if ($key === $needle) {
+                return true;
+            }
+
+            if (is_array($value)) {
+                if ($x = $this->keyExists($key, $value)) {
```

Comment:
The initialization of `$x` shouldn't be necessary.

Reply:
yes it was used to return the index initially, but forgot to remove it afterwards. will update that.